runtime/drivers/druid/druid_api_sql_driver.go

```diff
@@ -0,0 +1,354 @@
+						var l map[string]any
+						err := json.Unmarshal([]byte(v.(string)), &l)
+						if err != nil {
+							return nil, err
+						}
+						return l, nil
+					}
+				}
+			}
+
+			druidRows := &druidRows{
+				closer:       resp.Body,
+				dec:          dec,
+				jw:           jw,
+				columns:      columns,
+				types:        types,
+				transformers: transformers,
+			}
+			return druidRows, nil
+		}
+	}
+	return nil, jw.err
+}
+
+func toType(v any) string {
+	switch v.(type) {
+	case int:
+		return "INTEGER"
+	case float64:
+		return "DOUBLE"
+	case bool:
+		return "BOOLEAN"
+	default:
+		return "VARCHAR"
+	}
+}
+
+type druidRows struct {
+	closer       io.ReadCloser
+	dec          *json.Decoder
+	jw           *JSONWalker
+	columns      []string
+	types        []string
+	transformers []func(any) (any, error)
+}
+
+func (dr *druidRows) Columns() []string {
+	return dr.columns
+}
+
+func (dr *druidRows) Close() error {
+	return dr.closer.Close()
+}
+
+func (dr *druidRows) Next(dest []driver.Value) error {
+	if !dr.jw.hasMore() {
+		return io.EOF
+	} else if dr.jw.enterArray() {
+		values, err := dr.jw.arrayValues()
+		if err != nil {
+			return err
+		}
+
+		for i, v := range values {
+			v, err := dr.transformers[i](v)
+			if err != nil {
+				return err
+			}
+
+			dest[i] = v
+		}
+		return nil
+	}
+	return dr.jw.err
+}
+
+type stmt struct {
+	conn  *sqlConnection
+	query string
+}
+
+func (c *sqlConnection) Prepare(query string) (driver.Stmt, error) {
+	return &stmt{
+		query: query,
+		conn:  c,
+	}, nil
+}
+
+func (c *sqlConnection) Close() error {
+	return nil
+}
+
+func (c *sqlConnection) Begin() (driver.Tx, error) {
+	return nil, fmt.Errorf("unsupported")
+}
+
+func (s *stmt) Close() error {
+	return nil
+}
+
+func (s *stmt) NumInput() int {
+	return 0
+}
+
+type DruidQueryContext struct {
+	SQLQueryID string `json:"sqlQueryId"`
+}
+
+type DruidParameter struct {
+	Type  string `json:"type"`
+	Value any    `json:"value"`
+}
+
+type DruidRequest struct {
+	Query          string            `json:"query"`
+	Header         bool              `json:"header"`
+	SQLTypesHeader bool              `json:"sqlTypesHeader"`
+	ResultFormat   string            `json:"resultFormat"`
+	Parameters     []DruidParameter  `json:"parameters"`
+	Context        DruidQueryContext `json:"context"`
+}
+
+func druidRequest(query string, args []driver.NamedValue) *DruidRequest {
+	parameters := make([]DruidParameter, len(args))
+	for i, arg := range args {
+		parameters[i] = DruidParameter{
+			Type:  toType(arg.Value),
+			Value: arg.Value,
+		}
+	}
+	return &DruidRequest{
+		Query:          query,
+		Header:         true,
+		SQLTypesHeader: true,
+		ResultFormat:   "array",
+		Parameters:     parameters,
+		Context: DruidQueryContext{
+			SQLQueryID: uuid.New().String(),
+		},
+	}
+}
+
+func (s *stmt) Exec(args []driver.Value) (driver.Result, error) {
+	return nil, fmt.Errorf("unsupported")
+}
+
+type JSONWalker struct {
+	dec *json.Decoder
+	err error
+}
```

Comment:
The walker makes the logic pretty hard to follow – I'm guessing it's for performance? I wonder if there's a way to have Druid return NDJSON instead or some other kind of result pagination, so we could just use normal struct deserialization?

Reply:
ND arrays are possible:
```
["__time","aos","id","EXPR$3"]
["TIMESTAMP","OTHER","BIGINT","BOOLEAN"]
["2010-01-01T00:00:00.000Z","[{\"b\":\"gamma\",\"a\":1},{\"b\":\"beta\",\"a\":2}]",111,true]
```
Objects are possible but it's a waste of CPU and network:
```
{"__time":{"sqlType":"TIMESTAMP"},"aos":{"sqlType":"OTHER"},"id":{"sqlType":"BIGINT"},"EXPR$3":{"sqlType":"BOOLEAN"}}
{"__time":"2010-01-01T00:00:00.000Z","aos":"[{\"b\":\"gamma\",\"a\":1},{\"b\":\"beta\",\"a\":2}]","id":111,"EXPR$3":true}
```